File: cpp/time/rate_adjust.cpp
//-------------------------------------------------------------------------------------------
/*! \file    rate_adjust.cpp
    \brief   rate adjuster
    \author  Akihiko Yamaguchi, info@akihikoy.net
    \version 0.1
    \date    Aug.11, 2018

    g++ rate_adjust.cpp
    ./a.out
*/
//-------------------------------------------------------------------------------------------
#include <sys/time.h>
#include <unistd.h>

inline double GetCurrentTime(void)
{
  struct timeval time;
  gettimeofday (&time, NULL);
  return static_cast<double>(time.tv_sec) + static_cast<double>(time.tv_usec)*1.0e-6;
}

typedef double Duration;
typedef double Time;

// The code of the class Rate is from
// http://docs.ros.org/diamondback/api/rostime/html/classros_1_1Rate.html

class Rate
{
public:
  Rate(double frequency);

  bool sleep();

  void reset();

  Duration cycleTime();

  Duration expectedCycleTime() { return expected_cycle_time_; }

private:
  Time start_;
  Duration expected_cycle_time_, actual_cycle_time_;
};

Rate::Rate(double frequency)
: start_(GetCurrentTime())
, expected_cycle_time_(1.0 / frequency)
, actual_cycle_time_(0.0)
{}

bool Rate::sleep()
{
  Time expected_end = start_ + expected_cycle_time_;

  Time actual_end = GetCurrentTime();

  // detect backward jumps in time
  if (actual_end < start_)
  {
    expected_end = actual_end + expected_cycle_time_;
  }

  //calculate the time we'll sleep for
  Duration sleep_time = expected_end - actual_end;

  //set the actual amount of time the loop took in case the user wants to know
  actual_cycle_time_ = actual_end - start_;

  //make sure to reset our start time
  start_ = expected_end;

  //if we've taken too much time we won't sleep
  if(sleep_time <= Duration(0.0))
  {
    // if we've jumped forward in time, or the loop has taken more than a full extra
    // cycle, reset our cycle
    if (actual_end > expected_end + expected_cycle_time_)
    {
      start_ = actual_end;
    }
    return true;
  }

  return usleep(sleep_time*1.0e6);
}

void Rate::reset()
{
  start_ = GetCurrentTime();
}

Duration Rate::cycleTime()
{
  return actual_cycle_time_;
}


//-------------------------------------------------------------------------------------------
#include <iostream>
using namespace std;

int main(int argc, char**argv)
{
  Rate rate(5);

  for(int i(0);;++i)
  {
    cout<<i<<" "<<rate.cycleTime()<<endl;
    rate.sleep();
  }

  return 0;
}
//-------------------------------------------------------------------------------------------

File: python/time/rate_adjust.py
#!/usr/bin/python3
#\file    rate_adjust.py
#\brief   Rate adjuster.
#\author  Akihiko Yamaguchi, info@akihikoy.net
#\version 0.1
#\date    Jan.23, 2018

import time

'''Modified rospy.Rate with standard time.
https://docs.ros.org/api/rospy/html/rospy.timer-pysrc.html#Rate '''
class TRate(object):
  """
  Convenience class for sleeping in a loop at a specified rate
  """

  def __init__(self, hz, reset=False):
    """
    Constructor.
    @param hz: hz rate to determine sleeping
    @type  hz: float
    @param reset: if True, timer is reset when time moved backward. [default: False]
    @type  reset: bool
    """
    self.last_time = time.time()
    self.sleep_dur = 1.0/hz
    self._reset = reset

  def _remaining(self, curr_time):
    """
    Calculate the time remaining for rate to sleep.
    @param curr_time: current time
    @type  curr_time: L{Time}
    @return: time remaining
    @rtype: L{Time}
    """
    # detect time jumping backwards
    if self.last_time > curr_time:
      self.last_time = curr_time

    # calculate remaining time
    elapsed = curr_time - self.last_time
    return self.sleep_dur - elapsed

  def remaining(self):
    """
    Return the time remaining for rate to sleep.
    @return: time remaining
    @rtype: L{Time}
    """
    curr_time = time.time()
    return self._remaining(curr_time)

  def sleep(self):
    """
    Attempt sleep at the specified rate. sleep() takes into
    account the time elapsed since the last successful
    sleep().
    """
    curr_time = time.time()
    time.sleep(max(0.0, self._remaining(curr_time)))
    self.last_time = self.last_time + self.sleep_dur

    # detect time jumping forwards, as well as loops that are
    # inherently too slow
    if curr_time - self.last_time > self.sleep_dur * 2:
      self.last_time = curr_time



if __name__=='__main__':
  import random

  rate= TRate(2)
  t0= time.time()

  while True:
    print(('---------',time.time()-t0))
    t0= time.time()
    s= 0
    for i in range(int(600*random.random())):
      s+= i
      time.sleep(0.001)
    print(i)
    rate.sleep()


